File: src/populate.py
#!/usr/bin/env python
import boto3, json, os, os.path as op

LogFile = op.join( os.getenv('HOME'), '.cubelog' )
TableName = 'cubelog'

def main(cubelog=LogFile):
  with open(cubelog) as fh:
    ddb = boto3.resource('dynamodb')
    table = ddb.Table(TableName)
    for line in fh:
      fields = line[0:-1].split(maxsplit=3)
      timestamp = fields[0]
      puzzle = fields[1][1:-2]
      solve_time = fields[2]
      if len(fields) > 3:
        scramble = fields[3]
      else:
        scramble = None
      key = { 'puzzle': puzzle, 'timestamp': timestamp }
      try:
        item = table.get_item( Key = key )['Item']
      except KeyError:
        item = key
        key.update({'solve_time': solve_time, 'scramble': scramble })
        print(json.dumps(item))
        table.put_item( Item = item )

if __name__ == '__main__':
    main()

File: src/populate.pl
#!/usr/bin/env perl
use v5.16;
use JSON;
use Paws::DynamoDB;
use File::Spec::Functions;

our $LogFile = catfile $ENV{HOME}, '.cubelog';
our $TableName = 'cubelog';

sub main {
  my $puzzle = shift // '3x3x3';
  open my $fh, '<', $LogFile;
  my $ddb = Paws->service('DynamoDB', region => $ENV{AWS_DEFAULT_REGION} // 'us-east-1');
  while (chomp(my $line = <$fh>)) {
    my ($timestamp, $solve_time, $scramble)  = split ' ', $line, 3;
    chop($timestamp);
    my %key = ( 'puzzle' => { S => $puzzle }, 'timestamp' => { S => $timestamp } );
    my $item = $ddb->GetItem(TableName => $TableName, Key => \%key);
    unless ($item->Item) {
      my %item = (%key, solve_time => { S => $solve_time },
                 scramble => { S => $scramble} );
      $ddb->PutItem(TableName => $TableName, Item => \%item) or
        die "Failed to put item: $@\n";
    }
  }
}

main(@ARGV);
